### ISSUE TITLE
Module name prevents other modules in backports to import standard "os"

### ISSUE THREAD

**Issue by scriptorron** [pjdelport/backports.os#10]
The 64bit Anaconda 2.7 installs the modules:
- `backports`
- `backports.os` and
- `backports.shutil_which`.

The `backports.shutil_which` tries to import the standard `os` module with
`import os`
but in Python 2.7 the local `backports.os` module gets priority. Because the local `backports.os` does not implement everything from the standard `os` module an error will come up in `backports.shutil_which`:
```
  File "C:\Anaconda2_64bit\lib\site-packages\backports\shutil_which.py", line 12, in <module>
    def backport_which(cmd, mode=os.F_OK | os.X_OK, path=None):
AttributeError: 'module' object has no attribute 'F_OK'
```

The same issue will come up with every module in `backports` that tries to import the standard `os`.

For my opinion a module in backports that has the same name as a standard module should also provide all functionality of the standard module. An other module name should be used when only a part of the functionality is implemented (like `shutil_which` which reimplements only the `which` function of the `shutil` module).

One solution to work around this issue is to import everything from the standard `os` in the namespace of the `backports.os`. I have done this in my installation by inserting the following lines at the begin of `backports.os`:
 
```
from __future__ import unicode_literals
# begin of my bug fix
from __future__ import absolute_import
from os import *
# end of my bug fix

import sys
```

**Comment by PiDelport** [comment]
Hmm. I'm not sure about making `backports.os` import everything from `os`… isn't the more general convention for the backports packages to be explicit about what they backport, and only provide that?

My first impression is that the correct fix here would be to make `backports.shutil_which` declare `from __future__ import absolute_import` to avoid this problem.

**Comment by PiDelport** [comment]
I made a backports.shutil_which PR for this: https://github.com/minrk/backports.shutil_which/pull/3

@scriptorron: Does that change work for you?

**Comment by scriptorron** [comment]
@pjdelport: You wrote:

> isn't the more general convention for the backports packages to be explicit about what they backport, and only provide that?

I agree with that. But are you explicit about what you backport when you call it `os`? Other backports are called `functools_lru_cache` or `unittest_mock`. With such a name it is explicitely clear which function from what module is backported. When you just call it `os` I would expect that the whole `os` is backported.

I understand that it will be very difficult (or even impossible) to rename the file. Probably it is already used in too many other projects. FMO one solution would be to provide all `os` functionality in the backported `os`. Another solution would be to modify all other backports to use the global `os`. I do not know what way is the best.

**Comment by merwok** [comment]
The module is called backports.os, not os !

**Comment by PiDelport** [comment]
@scriptorron: Right, as @merwok said, the general intent (as I understand it) is that `backports.os` stays distinct from `os`.

Providing a reliable unified namespace is complicated, and makes it more complicated for client code to clearly choose between using the backport or not… so my hunch is still to keep the backported code separate like this.

**Comment by scriptorron** [comment]
@merwok, @PiDelport : Sorry, I do not agree with @merwok. Within the ```backports``` library the module is called ```os```! 

```backports``` is an open library where many people put their stuff in. Some of them need to import the standard ```os```. When they do this in the regular way with ```import os``` they do not get the standard ```os``` but the one in ```backports``` which does not provide all functionality of the standard module.

To reproduce the error please install ```backports.shutil_which``` and do
```
import backports
import backports.shutil_which
```
This is only an example. The error will come up with any other module in ```backports``` that tries to import the standard ```os```.

Why do you think do other peoble use explicit module names for their backported functions? I have in ```backports``` other modules with the following names:
```shutil_which``` (replaces ```which``` from module module ```shutil```)
```functools_lru_cache``` (replaces ```lru_cach``` from module ```functools```)
```ssl_match_hostname``` (replaces ```match_hostname``` from module ```ssl```)
```shutil_get_terminal_size``` (replaces ```get_terminal_size``` from module ```shutil```)

**Comment by merwok** [comment]
> The error will come up with any other module in backports that tries to import the standard os.

This should not happen if these other modules use absolute imports, which are a best practice with Python 2 and always enabled in Python 3.

**Comment by scriptorron** [comment]
@merwok: Please could you tell me what the best practice for importing the __standard__ ```os``` is? It seems that I did it wrong in the past.

**Comment by merwok** [comment]
Regular import rules:

In Python 3, a module `mypackage/util.py` with `import os` will import from the stdlib, not from `mypackage/os.py`.

For Python 2:
```
from __future__ import absolute_imports

import os   # stdlib!
from . import os  # sub-module
```

I checked the doc for namespace packages and I am not 100% sure whether these rules are different for them.

**Comment by scriptorron** [comment]
@merwok: I learned something! Thank you!!!!

What merwok proposes moves the ball to the other people who provide modules in ```backports```. Unfortunately not all of them are aware that they need to do an ```absolute_import``` to get the right ```os```.

At the end it does not matter to me how this issue is solved. I only want to be sure that Jupyter runs again. For me I found a solution by patching in the ```backport``` but I can imagine that other people will get frustrated.

**Comment by jaraco** [comment]
The danger with a package masking a stdlib package in a `backports` namespace is that all packages in that namespace need to be aware of this and ensure that `from __future__ import absolute_import` is present. Currently, [backports.pdb](/jaraco/backports.pdb) is probably incompatible with this package (due to the `import os`) on Python 2. Similarly, any backport that imports `pdb` would probably not work. Another backport that masks a stdlib module in the `backports` namespace is [configparser](/jaraco/configparser).

Probably this caveat should be brought up with the [backports project](https://bitbucket.org/brandon/backports) to be flagged up as something backports maintainers should check in their packages.

**Comment by PiDelport** [comment]
@jaraco: It's worth pointing out that this is a general Python 2 issue, though, and not anything specific to backports projects: all Python 2 code that isn't using `absolute_imports` has to be aware of this caveat already, and it will always be something that ultimately needs fixing in the code doing the importing.

**Comment by brandon-rhodes** [comment]
There might be a difference, though. In the general case, though, folks are writing their own packages and can control the problem by simply not adding modules with names like `pdb` — and they will, so to speak, suffer only for their own sins. But in this case, the author has no control over the adjacent names in the namespace their package lives in. And even worse, it will vary from one install to another. One set of users will remain happy with `backports.foo` while another set, mysteriously, will get errors the author has never heard of — because they've installed another `backports` module at their own sites.

Does that description capture the concern here?

**Comment by PiDelport** [comment]
Well, I think the important thing is that all Python 2 code (without absolute imports) is potentially buggy in this way. Renaming modules moves the problem around, but doesn't actually solve it: enabling absolute imports is the only way to reliably fix it, and is required to get the code more Python 3 compatible anyway.

Given that backports projects are generally intended to help code straddle Python 2 and 3, I'd consider it a bug in any backports module that isn't using absolute imports.

**Comment by PiDelport** [comment]
@scriptorron: Is there consensus that we should respond to this problem by enabling absolute imports in other affected modules, to close this issue?

**Comment by scriptorron** [comment]
For me it is okay. I opened this issue because Jupyter (a tool that is essential for me) was not working after an update and after an hour of debugging I found that the ```backports.shutil_which``` was importing the wrong ```os```. First I was thinking the ```backports.os``` is the reason for the trouble but now I have seen that the issue is much more complex and "philosophic". For me it does not matter how this issue is solved. I just wanted to avoid that other people get the same problem.
